### PR TITLE
Rewrite agent cache to be more memory efficient

### PR DESCRIPTION
**Pull request by azdagron**
- cache memory usage should be linear to the number of unique
selectors, registration entries, and subscribers
- reduce memory pressure by not cloning objects in or out of the cache
- use sync.Pool's for high-churn sets allocated during update/notification
- reduce sync code complexity by handling all of the "diff's" during a
single update call that can notify subscribers efficiently.
- fixes a few bugs in the old cache around propertly notifying subscribers
  when relevent information about registration entries is changed.

### CODE REVIEW

File: pkg/agent/manager/sync.go [hunk 1]
```diff
@@ -6,189 +6,148 @@ import (
...
+	// the values in `update` now belong to the cache. DO NOT MODIFY.
+	var csrs []csrRequest
+	var expiring int
+	m.cache.Update(update, func(entry *common.RegistrationEntry, svid *cache.X509SVID) {
+		var expiresAt time.Time
+		switch {
+		case svid == nil:
+			// no SVID
+		case len(svid.Chain) == 0:
+			// SVID has an empty chain. this is not expected to happen.
```
Comment (MarcosDY) [inline pkg/agent/manager/sync.go]: must not we do something here? notify in a log or somehitng than we are getting an svid without chain?

File: pkg/agent/manager/sync.go [hunk 2]
```diff
@@ -6,189 +6,148 @@ import (
...
+		log.Info("Renewing X509-SVID")
+		counter.AddLabel("spiffe_id", csr.SpiffeID)
 
-	update, err := m.client.FetchUpdates(ctx, &node.FetchX509SVIDRequest{Csrs: csrs})
-	if err != nil {
-		return nil, nil, err
-	}
+		// Skip CSR for the same SPIFFE ID... for now (see above TODO)
+		if _, ok := privateKeys[csr.SpiffeID]; ok {
+			continue
```
Comment (MarcosDY) [inline pkg/agent/manager/sync.go]: I know it will change in a future, but  maybe we can add a log to notify that private key will no be used, just to be able to debug what is happening because we "lost" some keys

File: pkg/agent/manager/cache/bundle_cache.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+// Changes returns the channel that is closed when a new value is available.
+func (b *BundleStream) Changes() chan struct{} {
+	return b.stream.Changes()
+}
+
+// Next advances this stream to the next state.
+// You should never call this unless Changes channel is closed.
+func (b *BundleStream) Next() map[string]*Bundle {
+	value, _ := b.stream.Next().(map[string]*Bundle)
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/bundle_cache.go]: I'd rather any changes to this Stream interface take place in a separate PR (this was just moved code)... the PR is big enough as is.

File: pkg/agent/manager/cache/cache.go [hunk 4]
```diff
@@ -2,357 +2,570 @@ package cache
...
+// selector sets and notifies subscribers when:
+//
+// 1) a registration entry related to the selectors:
+//   * is modified
+//   * has a new X509-SVID signed for it
+//   * federates with a federated bundle that is updated
+// 2) the trust bundle for the agent trust domain is updated
+//
+// When notified, the subscriber is given a WorkloadUpdate containing
+//.registration entries with their corresponding SVIDs and private keys along
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: 👍

File: pkg/agent/manager/cache/cache.go [hunk 5]
```diff
@@ -2,357 +2,570 @@ package cache
...
+// subscriptions, which can involve thousands of certificates, keys, bundles,
+// and registration entries, etc. The selector index itself is intended to be
+// scalable, but the objects themselves can take a considerable amount of
+// memory. For maximal safety, the objects should be cloned both coming in and
+// leaving the cache. However, during global updates (e.g. trust bundle is
+// updated for the agent trust domain) in particular, cloning all of the
+// revelant objects for each subscriber causes HUGE amounts of memory pressure
+// which adds non-trivial amounts of latency and causes a giant memory spike
+// that could OOM the agent on smaller VMs. For this reason, the cache is
+// presumed to own ALL data passing in and out of the cache. Produces MUST NOT
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: Fixed.

File: pkg/agent/manager/cache/cache.go [hunk 6]
```diff
@@ -2,357 +2,570 @@ package cache
...
+
+		selAdd, selRem := c.diffSelectors(set, entry, entry.regEntry, regEntry)
+		fedAdd, fedRem := c.diffFederatesWith(entry.regEntry, regEntry)
+
+		entry.regEntry = regEntry
+
+		// Now figure out if something related to the entry has changed so
+		// interested subscribers can be notified.
+		federatedBundlesChanged := fedAdd > 0 || fedRem > 0
+		for _, id := range regEntry.FederatesWith {
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: 👍

File: pkg/agent/manager/cache/cache.go [hunk 7]
```diff
@@ -2,357 +2,570 @@ package cache
...
+		c.BundleCache.Update(c.bundles)
 	}
-}
 
-func (c *cacheImpl) Bundle() *Bundle {
-	return c.Bundles()[c.trustDomain]
+	if trustDomainBundleChanged {
+		c.notifyAll()
+	} else {
+		c.notifySet(set)
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: Unfortunately not. There is probably some code that could be bypassed but teasing that out would make the code much more complicated. Even if we're going to end up not using the set, the code involved in calculating it also 1) updates the selector index, 2) logs changes to the debug log.

File: pkg/agent/manager/cache/cache.go [hunk 8]
```diff
@@ -2,357 +2,570 @@ package cache
...
+	index := c.getSelectorIndex(s)
+	index.entries[entry] = struct{}{}
 }
 
-func (c *cacheImpl) FetchEntry(entryID string) *Entry {
-	c.m.Lock()
-	defer c.m.Unlock()
-	if entry, found := c.cache[entryID]; found {
-		return entry
+func (c *Cache) delSelectorIndexEntry(s selector, entry *cacheEntry) {
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: 👍

File: pkg/agent/manager/cache/cache.go [hunk 9]
```diff
@@ -2,357 +2,570 @@ package cache
...
+	}
 
-	federatedBundles := make(map[string]*Bundle)
-	for _, entry := range entries {
+	// add in the bundles
+	for _, entry := range w.Entries {
 		for _, federatesWith := range entry.RegistrationEntry.FederatesWith {
-			if federatedBundle := bundles[federatesWith]; federatedBundle != nil {
-				federatedBundles[federatesWith] = federatedBundle
+			if federatedBundle := c.bundles[federatesWith]; federatedBundle != nil {
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: that's probably worth logging. good catch.

File: pkg/agent/manager/cache/cache.go [hunk 10]
```diff
@@ -2,357 +2,570 @@ package cache
...
-	return svid, ok
+func newCacheEntry() *cacheEntry {
+	return &cacheEntry{
+		subs: make(map[*subscriber]struct{}),
+	}
 }
 
-func (c *cacheImpl) SetJWTSVID(spiffeID string, audience []string, svid *client.JWTSVID) {
-	key := keyFromJWTSpiffeIDAndAudience(spiffeID, audience)
+type selectorIndex struct {
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: when subs _AND_ entries are empty. yes.

File: pkg/agent/manager/cache/cache_test.go [hunk 11]
```diff
@@ -1,395 +1,584 @@
...
-	"github.com/stretchr/testify/require"
 )
 
 var (
-	privateKey, _ = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
-	logger        logrus.FieldLogger
+	bundleV1      = bundleutil.BundleFromRootCA("spiffe://domain.test", &x509.Certificate{Raw: []byte{1}})
+	bundleV2      = bundleutil.BundleFromRootCA("spiffe://domain.test", &x509.Certificate{Raw: []byte{2}})
+	bundleV3      = bundleutil.BundleFromRootCA("spiffe://domain.test", &x509.Certificate{Raw: []byte{3}})
+	otherBundleV1 = bundleutil.BundleFromRootCA("spiffe://otherdomain.test", &x509.Certificate{Raw: []byte{2}})
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache_test.go]: it isn't important, but i'll give it a distinct value just to not cause doubt :)

File: pkg/agent/manager/cache/sets.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+func freeSelectorSet(set selectorSet) {
+	for k := range set {
+		delete(set, k)
+	}
+	selectorSetPool.Put(set)
+}
+
+func (set selectorSet) Merge(ss ...*common.Selector) {
+	for _, s := range ss {
+		set[selector{Type: s.Type, Value: s.Value}] = struct{}{}
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/sets.go]: practically speaking, no (since these are ultimately sourced from deserialized protobufs). there are lots of other places in the codebase that assume selectors in slices aren't nil, so if that is something we're worried about we have a wider audit to worry about.

File: pkg/agent/manager/sync.go (hunk 1 above)
Comment (azdagron) [reply pkg/agent/manager/sync.go]: sure, i can log.

File: pkg/agent/manager/sync.go (hunk 2 above)
Comment (azdagron) [reply pkg/agent/manager/sync.go]: sure, i will add a debug log.

File: pkg/agent/manager/cache/cache.go [hunk 13]
```diff
@@ -2,357 +2,582 @@ package cache
...
+		c.notifyAll()
+	} else {
+		c.notifySet(set)
+	}
 }
 
-func (c *cacheImpl) Bundles() map[string]*Bundle {
-	return c.bundles.Value().(map[string]*Bundle)
-}
+func (c *Cache) diffSelectors(set selectorSet, entry *cacheEntry, oldRegEntry, newRegEntry *common.RegistrationEntry) (added, removed int) {
```
[the remark this answers is not in the document]
Comment (evan2645) [reply pkg/agent/manager/cache/cache.go]: It is confusing that this is named `diffSelectors` but it also mutates the provided set, especially when `diffFederatesWith` does not have this behavior. We should either rename this function, or leave set mutation to the caller or a different (yet-to-be-defined) function

File: pkg/agent/manager/cache/cache.go [hunk 14]
```diff
@@ -2,357 +2,582 @@ package cache
...
-	h := sha1.New()
-
-	// duplicate and sort the audience slice before sorting
-	audience = append([]string(nil), audience...)
-	sort.Strings(audience)
-
-	io.WriteString(h, spiffeID)
-	for _, a := range audience {
-		io.WriteString(h, a)
+func makeEntry(entry *cacheEntry) Entry {
```
Comment (evan2645) [inline pkg/agent/manager/cache/cache.go]: We should qualify all of these `entry`s e.g. sortCacheEntries, makeCacheEntry, etc. Some of this logic is really hard to follow without a distinction

File: pkg/agent/manager/cache/bundle_cache.go [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+func NewBundleStream(stream observer.Stream) *BundleStream {
+	return &BundleStream{
+		stream: stream,
+	}
+}
+
+// Value returns the current value for this stream.
+func (b *BundleStream) Value() map[string]*Bundle {
+	value, _ := b.stream.Value().(map[string]*Bundle)
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/bundle_cache.go]: this was existing code that got moved but i'm happy to make that change

File: pkg/agent/manager/cache/cache.go [hunk 16]
```diff
@@ -2,357 +2,582 @@ package cache
...
+// Entries is only used by manager tests.
+// TODO: We should remove this and find a better way.
+func (c *Cache) Entries() []Entry {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	out := make([]Entry, 0, len(c.entries))
+	for _, entry := range c.entries {
+		if entry.svid == nil {
+			continue
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: The sync code  calls FetchX509SVID twice, once to learn about entries and the second to get some SVIDs signed. Between those two calls a cache entry won't have an SVID and shouldn't be returned from the cache. I'll leave a comment.

File: pkg/agent/manager/cache/cache.go [hunk 17]
```diff
@@ -2,357 +2,582 @@ package cache
...
-	bundles := map[string]*Bundle{
-		trustDomain: bundle,
+func (c *Cache) Update(update *CacheUpdate, checkSVID func(*common.RegistrationEntry, *X509SVID)) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// remove bundles that no longer exist and update bundle with changes,
+	// populating a "changed" set that we can check when processing
+	// registration entries to know if they need to spawn a notification.
+	// never remove the bundle for the trust domain.
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: Sure, I can split the comment. The "never remove the bundle for the trust domain" was a safety measure the previous cache implementation had in place to make sure that no matter what we had something to use to authorize the server (in case there was a bug where the server opted to not return the bundle). I'll try and clean up the language.

File: pkg/agent/manager/cache/cache.go [hunk 18]
```diff
@@ -2,357 +2,582 @@ package cache
...
+	bundleRemoved := false
+	for id := range c.bundles {
+		if _, ok := update.Bundles[id]; id != c.trustDomainID && !ok {
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: 👍

File: pkg/agent/manager/cache/cache.go [hunk 19]
```diff
@@ -2,357 +2,582 @@ package cache
...
-	// merge in changes
-	changed := false
-	for id, newBundle := range newBundles {
-		bundle := bundles[id]
-		if bundle == nil || !bundle.EqualTo(newBundle) {
-			bundles[id] = newBundle
-			changed = true
+
+	// add/update registration entries
+	for _, regEntry := range update.RegistrationEntries {
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: 👍

File: pkg/agent/manager/cache/cache.go [hunk 20]
```diff
@@ -2,357 +2,582 @@ package cache
...
+			if fedAdd > 0 {
+				log = log.WithField("fed_add", fedAdd)
+			}
+			if fedRem > 0 {
+				log = log.WithField("fed_rem", fedRem)
+			}
+			if entryExists {
+				log.Debug("Entry updated")
+			} else {
+				log.Debug("Entry created")
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: I added it because this refactor was huge and I wanted to log as much info as possible in case there were bugs introduced. It doesn't print the selectors or bundles themselves, but rather a count of the changes. Something like:

```
DEBU[0000] Entry created                                 entry_id=0a4eaff0-1ad8-4624-8d74-93c07fb1c9e6 sel_add=1 spiffe_id="spiffe://example.org/node" subsystem_name=cache
INFO[0000] Renewing X509-SVID                            spiffe_id="spiffe://example.org/node" subsystem_name=manager
DEBU[0000] Entry updated                                 entry_id=0a4eaff0-1ad8-4624-8d74-93c07fb1c9e6 spiffe_id="spiffe://example.org/node" subsystem_name=cache svid_updated=true
```

I'd like to keep it (or at least some version of it). If you are uncomfortable with the verbosity, i could log the ugliness under Trace level or something.... thoughts?

File: pkg/agent/manager/cache/jwt_cache.go [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.svids[key] = svid
+}
+
+func jwtSVIDKey(spiffeID string, audience []string) string {
+	h := sha1.New()
+
+	// duplicate and sort the audience slice before sorting
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/jwt_cache.go]: :+1

File: pkg/agent/manager/cache/cache.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (evan2645) [reply pkg/agent/manager/cache/cache.go]: Fair enough. There is no trace level IIRC so debug is fine

File: pkg/agent/manager/cache/subscriber.go [hunk 22]
```diff
@@ -3,140 +3,58 @@ package cache
...
-	defer s.m.Unlock()
-	s.sidMap[sub.sid] = sub
-
-	selSet := selector.NewSetFromRaw(sub.sel)
-	selPSet := selSet.Power()
-	for sel := range selPSet {
-		selStr := sel.String()
-		s.selMap[selStr] = append(s.selMap[selStr], sub.sid)
+		cache: cache,
+		set:   allocSelectorSet(selectors...),
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/subscriber.go]: nope, good catch. I'll clean it up after unsubscribing in `Finish()`.

File: pkg/agent/manager/sync.go [hunk 23]
```diff
@@ -6,189 +6,154 @@ import (
...
 	"github.com/spiffe/spire/pkg/common/util"
 	"github.com/spiffe/spire/proto/spire/api/node"
 	"github.com/spiffe/spire/proto/spire/common"
-	"github.com/zeebo/errs"
 )
 
+type csrRequest struct {
+	EntryID   string
+	SpiffeID  string
+	ExpiresAt time.Time
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/sync.go]: 👍

File: pkg/agent/manager/cache/cache.go [hunk 24]
```diff
@@ -2,357 +2,582 @@ package cache
...
+		}
+	}
 }
 
-func (c *cacheImpl) notifySubscribers(subs []*subscriber) {
-	if subs == nil {
-		return
+func (c *Cache) notifyAll() {
+	subs := c.allSubscribers()
+	defer freeSubscriberSet(subs)
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: i'm good with that

File: pkg/agent/manager/cache/cache.go (hunk 14 above)
Comment (azdagron) [reply pkg/agent/manager/cache/cache.go]: 👍
